# Homework 02-regression

## Preparing the dataset

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import mean_squared_error

In [10]:
url = "car_fuel_efficiency.csv"
df = pd.read_csv(url)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


In [17]:
df.describe

<bound method NDFrame.describe of       engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003

## Q1

In [55]:
# We can find which columns have missing values
print(df.isnull().sum())

# To many columns, we need to filter to only the one we need
print(f"\n'engine_displacement':", df['engine_displacement'].isnull().sum())
print(f"'horsepower':", df['horsepower'].isnull().sum())
print(f"'vehicle_weight':",df['vehicle_weight'].isnull().sum())
print(f"'model_year':", df['model_year'].isnull().sum())

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

'engine_displacement': 0
'horsepower': 708
'vehicle_weight': 0
'model_year': 0


In [56]:
# Then I want to have a ratio over how many missing value we have versus how many value we have in total.
missing = df['horsepower'].isna().sum()
total   = len(df['horsepower'])
ratio   = missing / total
print(f"\nMissing horsepower: {missing}/{total} ({ratio:.2%})")


Missing horsepower: 708/9704 (7.30%)


**Q1 : 'horsepower' is the one with missing values has we can he's not the only one in it.**

In [57]:
# We can also check for duplicates.
duplicates = df.duplicated()
print(f"Duplicated rows:",duplicates)

Duplicated rows: 0       False
1       False
2       False
3       False
4       False
        ...  
9699    False
9700    False
9701    False
9702    False
9703    False
Length: 9704, dtype: bool


## Q2

In [39]:
df.describe()

engine_displacement  num_cylinders   horsepower  vehicle_weight  \
count          9704.000000    9222.000000  8996.000000     9704.000000   
mean            199.708368       3.962481   149.657292     3001.280993   
std              49.455319       1.999323    29.879555      497.894860   
min              10.000000       0.000000    37.000000      952.681761   
25%             170.000000       3.000000   130.000000     2666.248985   
50%             200.000000       4.000000   149.000000     2993.226296   
75%             230.000000       5.000000   170.000000     3334.957039   
max             380.000000      13.000000   271.000000     4739.077089   

       acceleration   model_year    num_doors  fuel_efficiency_mpg  
count   8774.000000  9704.000000  9202.000000          9704.000000  
mean      15.021928  2011.484027    -0.006412            14.985243  
std        2.510339     6.659808     1.048162             2.556468  
min        6.000000  2000.000000    -4.000000             6.200971  
25%       13.300000  2006.000000    -1.000000            13.267459  
50%       15.000000  2012.000000     0.000000            15.006037  
75%       16.700000  2017.000000     1.000000            16.707965  
max       24.300000  2023.000000     4.000000            25.967222

In [40]:
df.describe(include='all')

engine_displacement  num_cylinders   horsepower  vehicle_weight  \
count           9704.000000    9222.000000  8996.000000     9704.000000   
unique                  NaN            NaN          NaN             NaN   
top                     NaN            NaN          NaN             NaN   
freq                    NaN            NaN          NaN             NaN   
mean             199.708368       3.962481   149.657292     3001.280993   
std               49.455319       1.999323    29.879555      497.894860   
min               10.000000       0.000000    37.000000      952.681761   
25%              170.000000       3.000000   130.000000     2666.248985   
50%              200.000000       4.000000   149.000000     2993.226296   
75%              230.000000       5.000000   170.000000     3334.957039   
max              380.000000      13.000000   271.000000     4739.077089   

        acceleration   model_year  origin fuel_type       drivetrain  \
count    8774.000000  9704.000000    9704      9704             9704   
unique           NaN          NaN       3         2                2   
top              NaN          NaN  Europe  Gasoline  All-wheel drive   
freq             NaN          NaN    3254      4898             4876   
mean       15.021928  2011.484027     NaN       NaN              NaN   
std         2.510339     6.659808     NaN       NaN              NaN   
min         6.000000  2000.000000     NaN       NaN              NaN   
25%        13.300000  2006.000000     NaN       NaN              NaN   
50%        15.000000  2012.000000     NaN       NaN              NaN   
75%        16.700000  2017.000000     NaN       NaN              NaN   
max        24.300000  2023.000000     NaN       NaN              NaN   

          num_doors  fuel_efficiency_mpg  
count   9202.000000          9704.000000  
unique          NaN                  NaN  
top             NaN                  NaN  
freq            NaN                  NaN  
mean      -0.006412            14.985243  
std        1.048162             2.556468  
min       -4.000000             6.200971  
25%       -1.000000            13.267459  
50%        0.000000            15.006037  
75%        1.000000            16.707965  
max        4.000000            25.967222

In [54]:
df['horsepower'].median()

np.float64(149.0)

**Q2 : We can already seen that in describre, the value of the median for 'horsepower', and we have confirmed his value**\\  
**Q2 : 149**

## Q3

In [58]:
# We need to split the data
from sklearn.model_selection import train_test_split

df_model = df[['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year', 'fuel_efficiency_mpg']].copy()
df_model = df_model.sample(frac=1, random_state=42)  # shuffle

n = len(df_model)
n_train = int(0.6 * n)
n_val   = int(0.2 * n)

df_train = df_model.iloc[:n_train]
df_val   = df_model.iloc[n_train:n_train+n_val]
df_test  = df_model.iloc[n_train+n_val:]

In [59]:
y_train = df_train.fuel_efficiency_mpg.values
y_val   = df_val.fuel_efficiency_mpg.values

X_train = df_train.drop(columns=['fuel_efficiency_mpg'])
X_val   = df_val.drop(columns=['fuel_efficiency_mpg'])

In [ ]:
# We saw previously that we have some missing values, then we have 2 options 
# Option A : Change the value by 0
X_train_0 = X_train.fillna(0)
X_val_0   = X_val.fillna(0)

# Option B : Using the mean
mean_hp = X_train['horsepower'].mean()
X_train_mean = X_train.fillna({'horsepower': mean_hp})
X_val_mean   = X_val.fillna({'horsepower': mean_hp})

In [61]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

def rmse(y, y_pred): 
    return np.sqrt(mean_squared_error(y, y_pred))

# (A) Imputation par 0
model_0 = LinearRegression()
model_0.fit(X_train_0, y_train)
rmse_0 = rmse(y_val, model_0.predict(X_val_0))

# (B) Imputation par moyenne
model_mean = LinearRegression()
model_mean.fit(X_train_mean, y_train)
rmse_mean = rmse(y_val, model_mean.predict(X_val_mean))

print(f"RMSE (fill=0): {rmse_0:.2f}")
print(f"RMSE (fill=mean): {rmse_mean:.2f}")

RMSE (fill=0): 0.52
RMSE (fill=mean): 0.46


``RMSE(mean)<RMSR(0)``, best option is to use the mean. In this cas lower is better.

**Q3 : `RMSE(mean)` is the answer

## Correlation between horsepower and fuel_efficiency_mpg 

In [63]:
corr = df[['horsepower', 'fuel_efficiency_mpg']].corr()
corr

horsepower  fuel_efficiency_mpg
horsepower             1.000000             0.122041
fuel_efficiency_mpg    0.122041             1.000000